### PR TITLE
make dict iterator use atomic operation

### PR DESCRIPTION
**Pull request by eldpswp99**
in #2915, there was a comment to use only fetch_add, not using load & store to maintain atomic operation.

since dict iterator has this load & store, change to use fetch_add.

### CODE REVIEW

File: vm/src/dictdatatype.rs [hunk 1]
```diff
@@ -488,6 +488,17 @@ impl<T: Clone> Dict<T> {
         }
     }
 
+    pub fn next_entry(&self, position: &AtomicCell<usize>) -> Option<(PyObjectRef, T)> {
```
Comment (coolreader18) [inline vm/src/dictdatatype.rs:491]: Maybe this could be named `next_entry_atomic`?

**Review by coolreader18 — APPROVED** [review]
Nice! This is definitely better.

File: vm/src/dictdatatype.rs (hunk 1 above)
Comment (eldpswp99) [reply vm/src/dictdatatype.rs:491]: Is it better to revert `_next_entry` to `next_entry` when I change the function name to `next_entry_atomic`?

File: vm/src/builtins/dict.rs [hunk 2]
```diff
@@ -718,12 +718,8 @@ macro_rules! dict_iterator {
                                 "dictionary changed size during iteration".to_owned(),
                             ));
                         }
-                        let mut position = zelf.position.load();
-                        match zelf.dict.entries.next_entry(&mut position) {
-                            Some((key, value)) => {
-                                zelf.position.store(position);
-                                Ok(($result_fn)(vm, key, value))
-                            }
+                        match zelf.dict.entries.next_entry_atomic(&zelf.position) {
```
Comment (youknowone) [inline vm/src/builtins/dict.rs:721]:
```suggestion
                        match zelf.dict.entries.next_entry(&zelf.position) {
```

File: vm/src/builtins/dict.rs [hunk 3]
```diff
@@ -782,7 +778,7 @@ macro_rules! dict_iterator {
                                 "dictionary changed size during iteration".to_owned(),
                             ));
                         }
-                        match zelf.dict.entries.next_entry_reversed(&zelf.position) {
+                        match zelf.dict.entries.next_entry_atomic_reversed(&zelf.position) {
```
Comment (youknowone) [inline vm/src/builtins/dict.rs:781]:
```suggestion
                        match zelf.dict.entries.next_entry_reversed(&zelf.position) {
```

File: vm/src/dictdatatype.rs [hunk 4]
```diff
@@ -488,7 +488,21 @@ impl<T: Clone> Dict<T> {
         }
     }
 
-    pub fn next_entry_reversed(&self, position: &AtomicCell<usize>) -> Option<(PyObjectRef, T)> {
+    pub fn next_entry_atomic(&self, position: &AtomicCell<usize>) -> Option<(PyObjectRef, T)> {
```
Comment (youknowone) [inline vm/src/dictdatatype.rs:491]:
```suggestion
    pub fn next_entry(&self, position: &AtomicCell<usize>) -> Option<(PyObjectRef, T)> {
```

File: vm/src/dictdatatype.rs [hunk 5]
```diff
@@ -488,7 +488,21 @@ impl<T: Clone> Dict<T> {
...
+        loop {
+            let position_usize = position.fetch_add(1);
+            let entry = inner.entries.get(position_usize)?;
+            if let Some(entry) = entry {
+                break Some((entry.key.clone(), entry.value.clone()));
+            }
+        }
+    }
+
+    pub fn next_entry_atomic_reversed(
```
Comment (youknowone) [inline vm/src/dictdatatype.rs:502]:
```suggestion
    pub fn next_entry_reversed(
```

**Pull request merged**